packages/shopping/src/services/refreshtoken.service.ts

```diff
@@ -0,0 +1,94 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/authentication
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+import {TokenService} from '@loopback/authentication';
+import {inject} from '@loopback/core';
+import {repository} from '@loopback/repository';
+import {HttpErrors} from '@loopback/rest';
+import {UserProfile, securityId} from '@loopback/security';
+import {UserRefreshtokenRepository} from '../repositories/user-refreshtoken.repository';
+import {RefreshtokenServiceBindings} from '../keys';
+
+export interface RefreshtokenService extends TokenService {
+  /**
+   * Verifies the validity of a token string and returns a user profile
+   *
+   * TODO(derdeka) move optional parameter userProfile to TokenService?
+   */
+  verifyToken(token: string, userProfile?: UserProfile): Promise<UserProfile>;
+  /**
+   * Revokes a given token (if supported by token system)
+   */
+  revokeToken(token: string, userProfile?: UserProfile): Promise<void>;
+}
+
+export class MyRefreshtokenService implements RefreshtokenService {
+  constructor(
+    @repository(UserRefreshtokenRepository)
+    public userRefreshtokenRepository: UserRefreshtokenRepository,
+    @inject(RefreshtokenServiceBindings.REFRESHTOKEN_ETERNAL_ALLOWED, {
+      optional: true,
+    })
+    private refreshtokenEternalAllowed: boolean = false,
+    @inject(RefreshtokenServiceBindings.REFRESHTOKEN_EXPIRES_IN, {
+      optional: true,
+    })
+    private refreshtokenExpiresIn: number = 60 * 60 * 24,
+  ) {}
+
+  async generateToken(userProfile: UserProfile): Promise<string> {
+    // TODO(derdeka) objectId as refreshtoken is a bad idea
+    const userRefreshtoken = await this.userRefreshtokenRepository.create({
+      creation: new Date(),
+      ttl: this.refreshtokenExpiresIn,
+      userId: userProfile[securityId],
+    });
+    return userRefreshtoken.id;
+  }
+
+  async verifyToken(
+    refreshtoken: string,
+    userProfile?: UserProfile,
+  ): Promise<UserProfile> {
+    try {
+      if (!userProfile || !userProfile[securityId]) {
+        throw new HttpErrors.Unauthorized('Invalid refreshToken');
+      }
+      const {creation, ttl} = await this.userRefreshtokenRepository.findById(
+        refreshtoken,
+        {
+          where: {
```

Comment:
usually findById should return only one result, any reason adding an additional where filter here?

Reply:
The idea is to check if the `refreshToken` is valid for / belongsTo the current user, validated by the authorization system. It's a kind of additional security check.

Reply:
I recently lerned from @raymondfeng that findById does not respect the where condition. I have to do the check differently.